contrib/babelfishpg_tsql/src/pl_gram.y

```diff
@@ -7218,6 +7224,31 @@ parse_datatype(const char *string, int location)
 	typeName = typeStringToTypeName(string);
 	typeName->names = rewrite_plain_name(typeName->names);
 	typenameTypeIdAndMod(NULL, typeName, &type_id, &typmod);
+	DeconstructQualifiedName(typeName->names, &schemaName, &dataTypeName);
+
+	/* in T-SQL, length-less (N)(VAR)CHAR's length is treated as 1 by default */
+	if (typmod == -1 && (is_tsql_varchar_or_char_datatype(type_id) || is_tsql_nchar_or_nvarchar_datatype(type_id) 
+									|| is_tsql_binary_or_varbinary_datatype(type_id)))
+		typmod = 1 + VARHDRSZ;
+	
+	/* for varchar/nvarchar/varbinary(MAX), set typmod back to -1 */
+	else if (typmod == TSQLMaxTypmod && is_tsql_datatype_with_max_scale_expr_allowed(type_id))
+		typmod = -1;
+	
+	else if (typmod > (8000 + VARHDRSZ) && (is_tsql_varchar_or_char_datatype(type_id) || is_tsql_binary_or_varbinary_datatype(type_id)))
```

Comment:
did we get this limits of 8000 and 4000 from SQL server? If yes, have we validated all test cases against SQL server